refactor(router): convert route attributes to plain objects

blackshadev opened the pull request:
# What

@aidan-casey made an alternative suggestion to removing the `Route` as abstract base. With this rework we also lose the `IsRoute` trait and all `Route` builder "verbs" (`Get`, `Post`, etc) become plain old php objects. No method, no logic, no magic. This will make it easier to use the `Route` interface. But it will make it "harder" to do logic in these Route builder "verbs" as you do not have any method to hook into, you should do everything in the constructor. 

Personally, I think this is as clean as it gets.

# DO NOT MERGE

Based on  #821 as PHPStan 2.0 is required, the old one will panic at the sight of property hooks in the interface

aidan-casey commented on .php-cs-fixer.dist.php:
```diff
@@ -9,6 +9,7 @@
     ])
     ->name('*.php')
     ->notName('*.cache.php')
+    ->notPath('Tempest/Router/src/Route.php') // phpcs doesn't yet support property hooks in interfaces
```
Simply commenting to say let's keep an eye on [this PR](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8312).

aidan-casey commented:
Cool PR!

brendt commented on src/Tempest/Router/src/Route.php:
```diff
@@ -8,10 +8,9 @@
 
-    /** @return class-string<HttpMiddleware>[] */
-    public function middleware(): array;
+    public array $middleware { get; }
```
Can't you keep the docblock?

blackshadev replied:
Yes I forgot, thanks!

brendt commented:
Nice!

blackshadev commented:
PHPStan seems drunk, can someone rerun it? It reports an error of unnecessary null safe, while on that line (and even in the file) no nullsafe operators are used...

aidan-casey commented:
@blackshadev I reran it from mobile, but same error. What file/line is it complaining about? It's not complaining locally?

blackshadev commented:
> @blackshadev I reran it from mobile, but same error. What file/line is it complaining about? It's not complaining locally?

It is not, but I found out why. Somebody other than me rebased this  branch from main , which contained the new error. I rebased the phpstan-v2 branch and added the error to the baseline. I will rebase this one to phpstan-v2 and it should be green.

brendt commented:
Very nice!

brendt commented:
Do we need to update the docs as well?

blackshadev commented:
> Do we need to update the docs as well?

A bit. I can do it tomorrow.

blackshadev commented:
> Do we need to update the docs as well?

updated [docs in this PR](https://github.com/tempestphp/tempest-docs/pull/48)
